src/i2c/controller.rs

```diff
@@ -413,55 +418,63 @@ impl<'d, T: Instance> I2c<'d, T, Async> {
 
         // send the start command
         let addr_rw = address << 1 | if read { 1 } else { 0 };
-        self.send_cmd(if self.is_hs { Cmd::StartHs } else { Cmd::Start }, addr_rw);
+        Self::send_cmd(if self.is_hs { Cmd::StartHs } else { Cmd::Start }, addr_rw);
 
         T::wait_cell()
             .wait_for(|| {
                 // enable interrupts
                 self.enable_tx_ints();
                 // if the command FIFO is empty, we're done sending start
-                self.is_tx_fifo_empty()
+                Self::is_tx_fifo_empty()
             })
             .await
             .map_err(|_| Error::Other)?;
 
-        self.status()
+        Self::status()
     }
 
     async fn async_stop(&mut self) -> Result<()> {
         // send the stop command
-        self.send_cmd(Cmd::Stop, 0);
+        Self::send_cmd(Cmd::Stop, 0);
 
         T::wait_cell()
             .wait_for(|| {
                 // enable interrupts
                 self.enable_tx_ints();
                 // if the command FIFO is empty, we're done sending stop
-                self.is_tx_fifo_empty()
+                Self::is_tx_fifo_empty()
             })
             .await
             .map_err(|_| Error::Other)?;
 
-        self.status()
+        Self::status()
     }
 
     async fn async_read_internal(&mut self, address: u8, read: &mut [u8], send_stop: SendStop) -> Result<()> {
         if read.is_empty() {
             return Err(Error::InvalidReadBufferLength);
         }
 
+        let on_drop = OnDrop::new(|| {
+            if Self::is_tx_fifo_full() {
```

Comment:
```suggestion
            if !Self::is_tx_fifo_empty() {
```

Should we reset whenever the fifo is not empty?